### PR TITLE
UPSTREAM: <carry>: openshift: Machine controller: drain node before machine deletion

### PR DESCRIPTION
**Pull request by ingvagabund**
**What this PR does / why we need it**:

The node draining code itself is imported from github.com/openshift/kubernetes-drain.
    
At the same time it's currently impossible to use the controller-runtime client for node draining due to missing Patch operation (kubernetes-sigs/controller-runtime#235). Thus, the machine controller needs to initialize kubeclient as well in order to implement the node draining logic. Once the Patch operation is implemented, the draining logic can be updated to replace kube client with controller runtime client.

Also, initialize event recorder to generate node draining event.

### CODE REVIEW

**Comment by ingvagabund** [comment]
/hold

We need to wait until all the tests in https://github.com/openshift/cluster-api-provider-aws/pull/174 pass (to verify the node draining is still functional).

**Comment by enxebre** [comment]
These are commits we want to carry on a rebase. Can you update the commits to follow `UPSTREAM: <carry>: openshift:` ?
https://github.com/openshift/origin/blob/master/docs/rebase.md#3-cherry-picks

File: pkg/controller/machine/controller.go [hunk 1]
```diff
@@ -145,6 +162,51 @@ func (r *ReconcileMachine) Reconcile(request reconcile.Request) (reconcile.Resul
...
 		}
 		klog.Infof("reconciling machine object %v triggers delete.", name)
+
+		// Drain node before deletion
+		// If a machine is not linked to a node, just delete the machine. Since a node
+		// can be unlinked from a machine when the node goes NotReady and is removed
+		// by cloud controller manager. In that case some machines would never get
+		// deleted without a manual intervention.
+		if _, exists := m.ObjectMeta.Annotations[ExcludeNodeDrainingAnnotation]; !exists && m.Status.NodeRef != nil {
+			if err := func() error {
```
Comment (frobware) [inline pkg/controller/machine/controller.go]: Why does this need to be a func?

Comment (ingvagabund) [reply pkg/controller/machine/controller.go]: To avoid repeated `return reconcile.Result{}, err` every time an error is returned.

Comment (bison) [reply pkg/controller/machine/controller.go]: Have to agree that this reads pretty weird. Can we just move it to an actual named function?

Comment (ingvagabund) [reply pkg/controller/machine/controller.go]: Done

**Pull request merged**